src/hnsw/utils.h

```diff
@@ -29,13 +31,19 @@ static inline void ldb_invariant(bool condition, const char *msg, ...)
 // search operations.
 // For this reason on hot codepaths we should avoid elog all together.
 // To print debug or test output on these hot codepaths, use ldb_dlog.
-// ldb_dlog has assert() semantics - it will be removed completely on release builds
-// so has no performance impact
-#ifdef LANTERNDB_DEBUG_LOGS
-#define ldb_dlog(elevel, ...) elog(elevel, errmsg_internal(__VA_ARGS__))
-#else
-#define ldb_dlog(elevel, ...) /*empty*/
-
-#endif  // LANTERNDB_DEBUG_LOGS
+#define ldb_dlog(...)                              \
+    do {                                           \
+        if(unlikely(ldb_is_test)) {                \
+            int   n, size = 200;                   \
+            char *buf = palloc(size);              \
+            n = snprintf(buf, size, __VA_ARGS__);  \
+            if(n >= size) {                        \
+                buf = repalloc(buf, n + 1);        \
+                snprintf(buf, n + 1, __VA_ARGS__); \
```

Comment:
Not sure what you are trying to do in the `if` block.

If the passed string interpolation is longer chan 200 chars, you repalloc to 201 chars  and re-interpolate the string.
Are you trying to make sure the string is NULL-terminated here? If so, you should pass `n` at line 42 and set the last char to `NULL`, explicitly. With the current approach, if the string interpolation result is more than 200 chars long, then the result at line 42 will not be null terminated either.

A shorter way to achieve the same would be to pass `size-1` at line 39 and do `palloc0()` at line 38 (`palloc` does not guarantee the returned memory will be zero-initialized)

I think it is much better to skip prepending LANTERN to messages and avoid the whole ordeal altogether.

Reply:
The goal is to allow handling debug logs of arbitrary length. `snprintf` should always append a null character I think. The if statement is to allocate extra space if the formatted output would be longer than 199 characters. `n` should hold the number of characters which would be generated for the given input, excluding the trailing null. 

It is definitely an ordeal just to log but I wanted to make it so that manually prepending `LANTERN` to debug logs wasn't required to surface them in regression tests

Reply:
> n should hold the number of characters which would be generated for the given input

I do not think this is true
see `snprintf` docs here: https://linux.die.net/man/3/snprintf

> **Return value**
> Upon successful return, these functions return the number of characters printed (excluding the null byte used to end output to strings).

Can you get rid of the whole prepending though? not worth it. You can add "LANTERN" to the string at callsite if you'd like